[docs] Include new documentation section on mining strategies (#5447)
debezium/debezium: documentation/modules/ROOT/pages/connectors/oracle.adoc

Instead, a redo entry holds a set of change vectors and a set of object identifiers that represent the tablespace, table, and columns related to these vectors.
In other words, you won't find the names of schemas, tables, or columns in these redo entries.

The {prodname} Oracle connector uses the xref:oracle-property-log-mining-strategy[`log.mining.strategy`] configuration property to control how Oracle LogMiner will handle the look-up of the object identifiers in the change vectors.
Certain strategies are more reliable than others for specific corner cases around schema changes; however, there are performance and overhead implications that should be considered.

==== Writing the data dictionary to redo logs

The default strategy is called `redo_log_catalog`, and this works by flushing a copy of the data dictionary to the redo logs immediately upon each redo log switch.
This is the most reliable strategy for tracking schema changes interwoven with data changes because Oracle LogMiner has a way to interpolate between the starting and ending data dictionary states across a series of change vectors.

> However, this mode is also the most experience because it requires several key steps to function.

Reviewer: ```suggestion
However, the `redo_log_catalog` mode is also the most expensive, because it requires several key steps to function.
```
Word choice?  Is this what you meant?

> First, the data dictionary must be flushed to the redo logs on each log switch.

Reviewer: ```suggestion
First, this mode requires the data dictionary to be flushed to the redo logs after every log switch.
```

> This can quickly consume valuable space for the database's archive log destination, while also generating substantially more archive logs than the database administrators may have previously accounted for.

Reviewer: ```suggestion
Flushing the logs after each switch can quickly consume valuable space in the archive log, and the high volume of archive logs might exceed the number that database administrators prepared for.
```

> It's imperative that administrators be aware if you intend to use this mode to make sure that the database is configured appropriately.

Reviewer: ```suggestion
If you intend to use this mode, coordinate with your database administrators to ensure that the database is configured appropriately.
```

[IMPORTANT]
====

> When using the `redo_log_catalog` mode, multiple {prodname} Oracle connectors should not be used to capture changes from the same logical database.

Reviewer: ```suggestion
If you configure the connector to use the `redo_log_catalog` mode, do not use multiple {prodname} Oracle connectors to capture changes from the same logical database.
```

====

==== Using the online catalog directly

> The next strategy mode is called `online_catalog`, and this works vastly different because the data dictionary is never flushed to the redo logs.

Reviewer: ```suggestion
The next strategy mode, `online_catalog`, works differently from the `redo_log_catalog` mode. 
When the strategy is set to `online_catalog`, the database never flushes the data dictionary to the redo logs.
```

> Instead, Oracle LogMiner will use the most current data dictionary state when doing comparisons.

Reviewer: ```suggestion
Instead, Oracle LogMiner always uses the most current data dictionary state to perform comparisons.
```

> This has the benefit that it has the least overhead, and operates the fastest; however, at the cost that interweaving schema changes with data changes can sometimes result in an event failure.

Reviewer: ```suggestion
By always using the current dictionary, and eliminating flushing to the redo logs, this strategy requires less overhead, and operates more efficiently.
However, these benefits are offset by the inability to parse interwoven schema changes and data changes.
As a result, this strategy can sometimes result in event failures.
```

Users who see the redo SQL refer to a table by a name like `OBJ# 123456` (where the number is its object identifier) or the columns as `COL1` or `COL2`; this means that a schema change was not performed in lock-step, and Oracle LogMiner was unable to reconstruct the SQL reliably for that change.

> When using this strategy, we recommend that the table's schema and its indices be as static and free from change as possible.

Reviewer: ```suggestion
When you configure the connector to use this strategy, to the extent that it's possible ensure that the table schema and its indices remain static and free from change.
```

> If you need to perform a schema change using the `online_catalog` mode, it's possible but you must follow these steps:

Reviewer: ```suggestion
If the {prodname} connector is configured to use the `online_catalog` mode, and you must apply a schema change, perform the following steps:
```

> 1. Wait for all existing data changes (DML) to be captured by the connector.

Reviewer: ```suggestion
1. Wait for the connector to capture all existing data changes (DML).
```

> 2. Perform the schema (DDL) change, waiting for the change to be captured by the connector.

Reviewer: ```suggestion
2. Perform the schema (DDL) change, and then wait for the connector to capture the change.
```

3. Resume data changes (DML) on the table.

> This lock-step procedure will guarantee that Oracle LogMiner will be able to safely reconstruct the SQL for all data changes.

Reviewer: ```suggestion
Following this procedure helps to ensure that Oracle LogMiner can safely reconstruct the SQL for all data changes.
```

==== Hybrid approach

> This is a new, experimental strategy that can be enabled by setting the strategy to `hybrid`.

Reviewer: ```suggestion
The `hybrid` log mining strategy is a new, experimental strategy.
```
Should this be conditionalized for community use? (and/or flagged as DP/TP for downstream?)

Author: It should :+1:

Author: Conditionalized.

> The goal of this strategy is to have all the performance and low overhead benefits of the _online catalog_ strategy but the reliability with schema changes from the _redo in catalog_ strategy, without any of the disadvantages from either strategy.

Reviewer: ```suggestion
The goal of this strategy is to provide the reliability of the `redo_log_catalog` strategy with the performance and low overhead of the `online catalog` strategy, without incurring the disadvantages of either strategy.
```

> This strategy works by primarily operating in the `online_catalog` mode, meaning that the {prodname} Oracle connector delegates the reconstruction first to Oracle LogMiner.

Reviewer: ```suggestion
The `hybrid` strategy works by primarily operating in the `online_catalog` mode, meaning that the {prodname} Oracle connector first delegates event reconstruction to Oracle LogMiner.
```

> If Oracle LogMiner is successful at the reconstruction of the SQL, the event is processed like normal as if using the `online_catalog` strategy.

Reviewer: ```suggestion
If Oracle LogMiner successfully reconstructs the SQL, {prodname} processes the event normally, as if it were configured to use the `online_catalog` strategy.
```

> When the {prodname} Oracle connector detects that Oracle LogMiner was unable to reconstruct the SQL, the connector will attempt to reconstruct the SQL directly by using the schema history for that table object.

Reviewer: ```suggestion
If the connector detects that Oracle LogMiner could not reconstruct the SQL, the connector attempts to reconstruct the SQL directly by using the schema history for that table object.
```

> If the SQL reconstruction fails by both Oracle LogMiner and the connector, only then will the connector report a failure.

Reviewer: ```suggestion
The connector reports a failure only if both Oracle LogMiner and the connector are unable to reconstruct the SQL.
```

}
----

[[oracle-logminer-mining-strategies]]
=== LogMiner Mining Strategies

Entries in the Oracle redo logs do not store the original SQL statements that users submit to make DML changes.
Instead, a redo entry holds a set of change vectors and a set of object identifiers that represent the tablespace, table, and columns related to these vectors.
In other words, redo log entries don't include the names of the schemas, tables, or columns affected by DML changes.

The {prodname} Oracle connector uses the xref:oracle-property-log-mining-strategy[`log.mining.strategy`] configuration property to control how Oracle LogMiner handles the lookup of the object identifiers in the change vectors.
In certain situations, one log mining strategy might prove more reliable than another with regard to schema changes.
However, before you choose a log mining strategy, it's important to consider the implications it might have on performance and overhead.

==== Writing the data dictionary to redo logs

> The default strategy is called `redo_log_catalog`.

Reviewer: ```suggestion
The default mining strategy is called `redo_log_catalog`.
```

Flushing the logs after each switch can quickly consume valuable space in the archive log, and the high volume of archive logs might exceed the number that database administrators prepared for.
If you intend to use this mode, coordinate with your database administrators to ensure that the database is configured appropriately.

[IMPORTANT]
====
If you configure the connector to use the `redo_log_catalog` mode, do not use multiple {prodname} Oracle connectors to capture changes from the same logical database.
====

==== Using the online catalog directly

The next strategy mode, `online_catalog`, works differently from the `redo_log_catalog` mode.
When the strategy is set to `online_catalog`, the database never flushes the data dictionary to the redo logs.
Instead, Oracle LogMiner always uses the most current data dictionary state to perform comparisons.
By always using the current dictionary, and eliminating flushing to the redo logs, this strategy requires less overhead, and operates more efficiently.
However, these benefits are offset by the inability to parse interwoven schema changes and data changes.
As a result, this strategy can sometimes result in event failures.

> Users who see the redo SQL refer to a table by a name like `OBJ# 123456` (where the number is its object identifier) or the columns as `COL1` or `COL2`; this means that a schema change was not performed in lock-step, and Oracle LogMiner was unable to reconstruct the SQL reliably for that change.

Reviewer: ```suggestion
If LogMiner was unable to reconstruct the SQL reliably after a schema change, check the redo logs for evidence . 
Look for references to tables with names like `OBJ# 123456` (where the number is the table's object identifier), or for columns with names like `COL1` or `COL2`. 
```

> When you configure the connector to use this strategy, to the extent that it's possible ensure that the table schema and its indices remain static and free from change.

Reviewer: ```suggestion
When you configure the connector to use the `online_catalog` strategy, take steps to ensure that the table schema and its indices remain static and free from change.
```
